src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,796 @@
+
+    OT_ASSERT(mIsRunning);
+
+    while (Get<NetworkData::Leader>().GetNextOnMeshPrefix(iterator, onMeshPrefixConfig) == OT_ERROR_NONE)
+    {
+        uint8_t newPrefixIndex;
+
+        if (!IsValidOmrPrefix(onMeshPrefixConfig.GetPrefix()) || !onMeshPrefixConfig.mDefaultRoute ||
+            !onMeshPrefixConfig.mSlaac || onMeshPrefixConfig.mDp)
+        {
+            continue;
+        }
+
+        newPrefixIndex = 0;
+        while (newPrefixIndex < newOmrPrefixNum && onMeshPrefixConfig.GetPrefix() != aNewOmrPrefixes[newPrefixIndex])
+        {
+            ++newPrefixIndex;
+        }
+
+        if (newPrefixIndex != newOmrPrefixNum)
+        {
+            // Ignore duplicate prefixes.
+            continue;
+        }
+
+        if (newOmrPrefixNum >= aMaxOmrPrefixNum)
+        {
+            otLogWarnBr("EvaluateOmrPrefix: too many OMR prefixes, ignoring prefix %s",
+                        onMeshPrefixConfig.GetPrefix().ToString().AsCString());
+            continue;
+        }
+
+        aNewOmrPrefixes[newOmrPrefixNum] = onMeshPrefixConfig.GetPrefix();
+        if (smallestOmrPrefix == nullptr || IsPrefixSmallerThan(onMeshPrefixConfig.GetPrefix(), *smallestOmrPrefix))
+        {
+            smallestOmrPrefix = &aNewOmrPrefixes[newOmrPrefixNum];
+        }
+        if (aNewOmrPrefixes[newOmrPrefixNum] == mLocalOmrPrefix)
+        {
+            publishedLocalOmrPrefix = &aNewOmrPrefixes[newOmrPrefixNum];
+        }
+        ++newOmrPrefixNum;
+    }
+
+    // Decide if we need to add or remove my local OMR prefix.
+    if (newOmrPrefixNum == 0)
+    {
+        otLogInfoBr("EvaluateOmrPrefix: no valid OMR prefixes found in Thread network");
+        if (PublishLocalOmrPrefix() == OT_ERROR_NONE)
+        {
+            aNewOmrPrefixes[newOmrPrefixNum++] = mLocalOmrPrefix;
+        }
+
+        // The `newOmrPrefixNum` is zero when we failed to publish the local OMR prefix.
+    }
+    else
+    {
+        if (publishedLocalOmrPrefix != nullptr && smallestOmrPrefix != publishedLocalOmrPrefix)
+        {
+            otLogInfoBr("EvaluateOmrPrefix: there is already a smaller OMR prefix %s in the Thread network",
+                        smallestOmrPrefix->ToString().AsCString());
+            UnpublishLocalOmrPrefix();
+
+            // Remove the local OMR prefix from the list by overwriting it with the last one.
+            *publishedLocalOmrPrefix = aNewOmrPrefixes[--newOmrPrefixNum];
+        }
+    }
+
+    return newOmrPrefixNum;
+}
+
+otError RoutingManager::PublishLocalOmrPrefix(void)
+{
+    otError                         error = OT_ERROR_NONE;
+    NetworkData::OnMeshPrefixConfig omrPrefixConfig;
+
+    OT_ASSERT(mIsRunning);
+
+    omrPrefixConfig.Clear();
+    omrPrefixConfig.mPrefix       = mLocalOmrPrefix;
+    omrPrefixConfig.mStable       = true;
+    omrPrefixConfig.mSlaac        = true;
+    omrPrefixConfig.mPreferred    = true;
+    omrPrefixConfig.mOnMesh       = true;
+    omrPrefixConfig.mDefaultRoute = true;
```

Comment:
Not always using default route is a future enhancement?

Reply:
Without setting it as default route, we will not be able to handle the multiple Thread Networks case unless we handle RIO sent by BRs in other Thread Networks and adds external routes for them. I am preparing the PR to add specific routes for current BR and "external" BRs, but it will be a significant change (around 1000 if include tests). I think it should be easier to review the change in a separate PR. thoughts?